Foundation/NSURLSession/NSURLSessionTask.swift

```diff
@@ -202,8 +202,21 @@ open class URLSessionTask : NSObject, NSCopying {
      * cases, the task may signal other work before it acknowledges the
      * cancelation.  -cancel may be sent to a task that has been suspended.
      */
-    open func cancel() { NSUnimplemented() }
-    
+    open func cancel() {
+        workQueue.sync {
+            guard self.state == .running || self.state == .suspended else { return }
+            self.state = .canceling
+            self.workQueue.async {
+                guard let s = self.session as? URLSession else { fatalError() }
+                guard let delegate  = s.delegate as? URLSessionTaskDelegate else { fatalError() }
+                s.delegateQueue.addOperation {
+                    delegate.urlSession(s, task: self, didCompleteWithError: NSError(domain: NSURLErrorDomain, code: NSURLErrorCancelled, userInfo: nil))
+                    self.internalState = .taskCompleted
+                }
```

Comment:
Wouldn't a call of `completeTask(withError:)` cover all situations when there is a need to perform a task completion?